crates/turbo-tasks-fs/src/lib.rs

```diff
@@ -1089,6 +1090,62 @@ impl FileSystemPathVc {
     pub async fn is_inside_or_equal(self, other: FileSystemPathVc) -> Result<BoolVc> {
         Ok(BoolVc::cell(self.await?.is_inside_or_equal(&*other.await?)))
     }
+
+    /// Creates a new [`FileSystemPathVc`] like `self` but with the given
+    /// extension.
+    #[turbo_tasks::function]
+    pub async fn with_extension(self, extension: &str) -> Result<FileSystemPathVc> {
+        let this = self.await?;
+        let path_without_extension =
+            if let Some((path_without_extension, old_extension)) = this.path.rsplit_once('.') {
+                if old_extension.contains('/') {
+                    this.path.as_str()
+                } else {
+                    path_without_extension
+                }
+            } else {
+                this.path.as_str()
+            };
+        Ok(Self::new_normalized(
+            this.fs,
+            // Like `Path::with_extension` and `PathBuf::set_extension`, if the extension is empty,
+            // we remove the extension altogether.
+            match extension.is_empty() {
+                true => format!("{path_without_extension}"),
+                false => format!("{path_without_extension}.{extension}"),
+            },
+        ))
+    }
+
+    /// Extracts the stem (non-extension) portion of self.file_name.
+    ///
+    /// The stem is:
+    ///
+    /// * [`None`], if there is no file name;
+    /// * The entire file name if there is no embedded `.`;
+    /// * The entire file name if the file name begins with `.` and has no other
+    ///   `.`s within;
+    /// * Otherwise, the portion of the file name before the final `.`
+    #[turbo_tasks::function]
+    pub async fn file_stem(self) -> Result<OptionStringVc> {
+        let this = self.await?;
+        if this.path.is_empty() {
+            return Ok(OptionStringVc::cell(None));
+        }
+        let file_name = if let Some((_prefix_path, file_name)) = this.path.rsplit_once('/') {
```

Comment:
Nit: A lot of methods deal with manipulating the `dirname`, `stem`, and `ext`, and have to rework this `rsplit('.')` (but only if there's no `/` in the extension). Can we rework this to a reusable method that returns a `(dirname, stem, ext)` tuple?